File: llmcalc/pricing_parser.py
"""Deterministic per-entry pricing decoding with structured diagnostics."""

from __future__ import annotations

import re
from collections.abc import Mapping
from decimal import Decimal, InvalidOperation
from typing import Any

from pydantic import ValidationError

from llmcalc.diagnostics import DiagnosticAction, DiagnosticSeverity, PricingDiagnostic
from llmcalc.models import ModelPricing, RawModelPricing
from llmcalc.pricing_tiers import RATE_FIELDS, parse_query_tiers, parse_tiers

_CORE_RATE_FIELDS = {
    "input_cost_per_token",
    "output_cost_per_token",
    "prompt_cost_per_token",
    "completion_cost_per_token",
    "input_cost_per_million_tokens",
    "output_cost_per_million_tokens",
    "input_cost_per_token_cache_hit",
    "input_cost_per_query",
    "google_maps_grounding_cost_per_query",
    "regional_processing_uplift_multiplier_us",
    "regional_processing_uplift_multiplier_eu",
    "regional_endpoint_uplift_multiplier",
    *RATE_FIELDS.values(),
}
_MODE_RATE = re.compile(
    r"^(?:" + "|".join(re.escape(field) for field in RATE_FIELDS.values()) + r")_"
    r"(?:batches|priority|flex)$"
)
_THRESHOLD_RATE = re.compile(
    r"^(?:input_cost_per_token|output_cost_per_token|cache_read_input_token_cost|"
    r"cache_creation_input_token_cost|cache_creation_input_token_cost_above_1hr)"
    r"_above_\d+k?_tokens(?:_(?:batches|priority|flex))?$"
)
_METADATA_FIELDS = ("provider", "litellm_provider", "currency")
_SEARCH_CONTEXT_KEYS = {
    "search_context_size_low",
    "search_context_size_medium",
    "search_context_size_high",
}


def _diagnostic(
    model: str,
    severity: DiagnosticSeverity,
    code: str,
    action: DiagnosticAction,
    path: tuple[str | int, ...],
    message: str,
) -> PricingDiagnostic:
    return PricingDiagnostic(
        model=model,
        severity=severity,
        code=code,
        action=action,
        path=path,
        message=message,
    )


def _valid_decimal(value: Any) -> bool:
    if value is None or isinstance(value, bool):
        return False
    try:
        parsed = Decimal(str(value).strip())
    except (InvalidOperation, ValueError):
        return False
    return parsed.is_finite() and parsed >= Decimal("0")


def _valid_multiplier(value: Any) -> bool:
    if not _valid_decimal(value):
        return False
    return Decimal(str(value).strip()) > Decimal("0")


def _is_supported_rate(key: str) -> bool:
    return key in _CORE_RATE_FIELDS or _MODE_RATE.fullmatch(key) is not None


def _has_other_supported_pricing(raw: Mapping[str, Any]) -> bool:
    if parse_query_tiers(raw.get("tiered_pricing")):
        return True
    for key, value in raw.items():
        if key == "tiered_pricing":
            continue
        if (_is_supported_rate(key) or _THRESHOLD_RATE.fullmatch(key)) and _valid_decimal(value):
            return True
        if (
            key == "search_context_cost_per_query"
            and isinstance(value, Mapping)
            and any(
                context in _SEARCH_CONTEXT_KEYS and _valid_decimal(rate)
                for context, rate in value.items()
            )
        ):
            return True
    return False


def _unsupported_cost_field(key: str) -> bool:
    return (
        "cost" in key
        and key not in {"search_context_cost_per_query"}
        and not _is_supported_rate(key)
        and _THRESHOLD_RATE.fullmatch(key) is None
    )


def inspect_model_pricing(
    model: str,
    model_data: Mapping[str, Any],
    default_currency: str,
) -> tuple[ModelPricing | None, tuple[PricingDiagnostic, ...]]:
    diagnostics: list[PricingDiagnostic] = []
    sanitized = dict(model_data)

    for field in _METADATA_FIELDS:
        value = sanitized.get(field)
        if value is not None and not isinstance(value, str):
            return None, (
                _diagnostic(
                    model,
                    "error",
                    "invalid_metadata",
                    "skipped_entry",
                    (field,),
                    "pricing metadata must use the expected type",
                ),
            )

    last_updated = sanitized.get("last_updated")
    if last_updated is not None and not isinstance(last_updated, str):
        diagnostics.append(
            _diagnostic(
                model,
                "warning",
                "invalid_metadata",
                "ignored_field",
                ("last_updated",),
                "last_updated must be a string",
            )
        )
        sanitized.pop("last_updated")

    provider_specific = sanitized.get("provider_specific_entry")
    if isinstance(provider_specific, Mapping):
        cleaned_provider_specific = dict(provider_specific)
        for field in ("us", "fast"):
            value = cleaned_provider_specific.get(field)
            if value is not None and not _valid_multiplier(value):
                diagnostics.append(
                    _diagnostic(
                        model,
                        "warning",
                        "invalid_rate",
                        "ignored_field",
                        ("provider_specific_entry", field),
                        "provider pricing multiplier must be a finite positive decimal",
                    )
                )
                cleaned_provider_specific.pop(field)
        sanitized["provider_specific_entry"] = cleaned_provider_specific

    for key, value in tuple(sanitized.items()):
        if _THRESHOLD_RATE.fullmatch(key) is not None and not _valid_decimal(value):
            diagnostics.append(
                _diagnostic(
                    model,
                    "warning",
                    "invalid_threshold_rate",
                    "ignored_field",
                    (key,),
                    "long-context pricing rate is malformed",
                )
            )
            sanitized.pop(key)
        elif _is_supported_rate(key) and not _valid_decimal(value):
            return None, (
                *diagnostics,
                _diagnostic(
                    model,
                    "error",
                    "invalid_rate",
                    "skipped_entry",
                    (key,),
                    "supported pricing rate must be a finite non-negative decimal",
                ),
            )

    raw_tiers = sanitized.get("tiered_pricing")
    has_range_tiers = isinstance(raw_tiers, (list, tuple)) and any(
        isinstance(entry, Mapping) and "range" in entry for entry in raw_tiers
    )
    has_query_tiers = isinstance(raw_tiers, (list, tuple)) and any(
        isinstance(entry, Mapping) and "max_results_range" in entry
        for entry in raw_tiers
    )
    invalid_tiers = (has_range_tiers and not parse_tiers(raw_tiers)) or (
        has_query_tiers and not parse_query_tiers(raw_tiers)
    )
    if invalid_tiers:
        other_pricing = _has_other_supported_pricing(sanitized)
        diagnostic = _diagnostic(
            model,
            "warning" if other_pricing else "error",
            "invalid_tier_schedule",
            "ignored_field" if other_pricing else "skipped_entry",
            ("tiered_pricing",),
            "request-tier pricing schedule is malformed",
        )
        if not other_pricing:
            return None, (*diagnostics, diagnostic)
        diagnostics.append(diagnostic)
        sanitized.pop("tiered_pricing")

    try:
        raw_model = RawModelPricing.model_validate(sanitized)
        pricing = raw_model.to_model_pricing(model, default_currency=default_currency)
    except ValidationError:
        return None, (
            *diagnostics,
            _diagnostic(
                model,
                "error",
                "invalid_metadata",
                "skipped_entry",
                (),
                "pricing entry does not match the supported schema",
            ),
        )
    except ValueError as error:
        if str(error) != "Missing supported pricing fields":
            raise
        has_cost_field = any("cost" in key for key in sanitized)
        return None, (
            *diagnostics,
            _diagnostic(
                model,
                "info",
                "unsupported_pricing_unit" if has_cost_field else "missing_supported_pricing",
                "skipped_entry",
                (),
                (
                    "entry uses pricing units that llmcalc does not support"
                    if has_cost_field
                    else "entry has no supported pricing fields"
                ),
            ),
        )
    for key in sanitized:
        if _unsupported_cost_field(key):
            diagnostics.append(
                _diagnostic(
                    model,
                    "info",
                    "unsupported_pricing_unit",
                    "ignored_field",
                    (key,),
                    "pricing field uses a billing unit that llmcalc does not support",
                )
            )
    return pricing, tuple(diagnostics)

File: js/src/pricing-parser.ts
import { Decimal } from "./decimal.js";
import {
  type DiagnosticAction,
  type DiagnosticSeverity,
  type PricingDiagnostic
} from "./diagnostics.js";
import { type ModelPricing, toModelPricing } from "./models.js";
import { RATE_FIELDS, parseQueryTiers, parseTiers } from "./pricing-tiers.js";

const CORE_RATE_FIELDS = new Set([
  "input_cost_per_token",
  "output_cost_per_token",
  "prompt_cost_per_token",
  "completion_cost_per_token",
  "input_cost_per_million_tokens",
  "output_cost_per_million_tokens",
  "input_cost_per_token_cache_hit",
  "input_cost_per_query",
  "google_maps_grounding_cost_per_query",
  "regional_processing_uplift_multiplier_us",
  "regional_processing_uplift_multiplier_eu",
  "regional_endpoint_uplift_multiplier",
  ...Object.values(RATE_FIELDS)
]);
const MODE_RATE = new RegExp(
  `^(?:${Object.values(RATE_FIELDS).join("|")})_(?:batches|priority|flex)$`
);
const THRESHOLD_RATE = new RegExp(
  "^(?:input_cost_per_token|output_cost_per_token|cache_read_input_token_cost|" +
    "cache_creation_input_token_cost|cache_creation_input_token_cost_above_1hr)" +
    "_above_\\d+k?_tokens(?:_(?:batches|priority|flex))?$"
);
const METADATA_FIELDS = ["provider", "litellm_provider", "currency"] as const;
const SEARCH_CONTEXT_KEYS = new Set([
  "search_context_size_low",
  "search_context_size_medium",
  "search_context_size_high"
]);

export interface ModelPricingInspection {
  pricing: ModelPricing | null;
  diagnostics: PricingDiagnostic[];
}

function diagnostic(
  model: string,
  severity: DiagnosticSeverity,
  code: string,
  action: DiagnosticAction,
  path: Array<string | number>,
  message: string
): PricingDiagnostic {
  return { model, severity, code, action, path, message };
}

function validDecimal(value: unknown): boolean {
  if (value === null || value === undefined || typeof value === "boolean") {
    return false;
  }
  if (typeof value !== "number" && typeof value !== "string") {
    return false;
  }
  try {
    const parsed = new Decimal(String(value).trim());
    return parsed.isFinite() && parsed.greaterThanOrEqualTo(0);
  } catch {
    return false;
  }
}

function validMultiplier(value: unknown): boolean {
  return validDecimal(value) && new Decimal(String(value).trim()).greaterThan(0);
}

function isSupportedRate(key: string): boolean {
  return CORE_RATE_FIELDS.has(key) || MODE_RATE.test(key);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasOtherSupportedPricing(raw: Record<string, unknown>): boolean {
  if (parseQueryTiers(raw.tiered_pricing).length > 0) {
    return true;
  }
  for (const [key, value] of Object.entries(raw)) {
    if (key === "tiered_pricing") {
      continue;
    }
    if ((isSupportedRate(key) || THRESHOLD_RATE.test(key)) && validDecimal(value)) {
      return true;
    }
    if (
      key === "search_context_cost_per_query" &&
      isRecord(value) &&
      Object.entries(value).some(
        ([context, rate]) => SEARCH_CONTEXT_KEYS.has(context) && validDecimal(rate)
      )
    ) {
      return true;
    }
  }
  return false;
}

function unsupportedCostField(key: string): boolean {
  return key.includes("cost") &&
    key !== "search_context_cost_per_query" &&
    !isSupportedRate(key) &&
    !THRESHOLD_RATE.test(key);
}

function matchesRawSchema(raw: Record<string, unknown>): boolean {
  const tiers = raw.tiered_pricing;
  if (tiers !== null && tiers !== undefined && !Array.isArray(tiers)) {
    return false;
  }

  const searchCosts = raw.search_context_cost_per_query;
  if (searchCosts !== null && searchCosts !== undefined) {
    if (!isRecord(searchCosts) || !Object.values(searchCosts).every(validDecimal)) {
      return false;
    }
  }

  const providerSpecific = raw.provider_specific_entry;
  if (providerSpecific !== null && providerSpecific !== undefined && !isRecord(providerSpecific)) {
    return false;
  }

  const billingUnit = raw.web_search_billing_unit;
  if (
    billingUnit !== null &&
    billingUnit !== undefined &&
    billingUnit !== "per_prompt" &&
    billingUnit !== "per_query"
  ) {
    return false;
  }

  for (const region of ["us", "eu"] as const) {
    const value = raw[`regional_processing_uplift_multiplier_${region}`];
    if (value !== null && value !== undefined && validDecimal(value)) {
      if (new Decimal(String(value).trim()).lessThanOrEqualTo(0)) {
        return false;
      }
    }
  }
  return true;
}

function schemaDiagnostic(model: string): PricingDiagnostic {
  return diagnostic(
    model,
    "error",
    "invalid_metadata",
    "skipped_entry",
    [],
    "pricing entry does not match the supported schema"
  );
}

export function inspectModelPricing(
  model: string,
  modelData: Record<string, unknown>,
  defaultCurrency: string
): ModelPricingInspection {
  const diagnostics: PricingDiagnostic[] = [];
  const sanitized = { ...modelData };

  for (const field of METADATA_FIELDS) {
    const value = sanitized[field];
    if (value !== null && value !== undefined && typeof value !== "string") {
      return {
        pricing: null,
        diagnostics: [
          diagnostic(
            model,
            "error",
            "invalid_metadata",
            "skipped_entry",
            [field],
            "pricing metadata must use the expected type"
          )
        ]
      };
    }
  }

  const lastUpdated = sanitized.last_updated;
  if (lastUpdated !== null && lastUpdated !== undefined && typeof lastUpdated !== "string") {
    diagnostics.push(
      diagnostic(
        model,
        "warning",
        "invalid_metadata",
        "ignored_field",
        ["last_updated"],
        "last_updated must be a string"
      )
    );
    delete sanitized.last_updated;
  }

  const providerSpecific = sanitized.provider_specific_entry;
  if (isRecord(providerSpecific)) {
    const cleaned = { ...providerSpecific };
    for (const field of ["us", "fast"] as const) {
      const value = cleaned[field];
      if (value !== null && value !== undefined && !validMultiplier(value)) {
        diagnostics.push(
          diagnostic(
            model,
            "warning",
            "invalid_rate",
            "ignored_field",
            ["provider_specific_entry", field],
            "provider pricing multiplier must be a finite positive decimal"
          )
        );
        delete cleaned[field];
      }
    }
    sanitized.provider_specific_entry = cleaned;
  }

  for (const [key, value] of Object.entries(sanitized)) {
    if (THRESHOLD_RATE.test(key) && !validDecimal(value)) {
      diagnostics.push(
        diagnostic(
          model,
          "warning",
          "invalid_threshold_rate",
          "ignored_field",
          [key],
          "long-context pricing rate is malformed"
        )
      );
      delete sanitized[key];
    } else if (isSupportedRate(key) && !validDecimal(value)) {
      return {
        pricing: null,
        diagnostics: [
          ...diagnostics,
          diagnostic(
            model,
            "error",
            "invalid_rate",
            "skipped_entry",
            [key],
            "supported pricing rate must be a finite non-negative decimal"
          )
        ]
      };
    }
  }

  const rawTiers = sanitized.tiered_pricing;
  const hasRangeTiers =
    Array.isArray(rawTiers) && rawTiers.some((entry) => isRecord(entry) && "range" in entry);
  const hasQueryTiers =
    Array.isArray(rawTiers) &&
    rawTiers.some((entry) => isRecord(entry) && "max_results_range" in entry);
  const invalidTiers =
    (hasRangeTiers && parseTiers(rawTiers).length === 0) ||
    (hasQueryTiers && parseQueryTiers(rawTiers).length === 0);
  if (invalidTiers) {
    const otherPricing = hasOtherSupportedPricing(sanitized);
    const invalidTier = diagnostic(
      model,
      otherPricing ? "warning" : "error",
      "invalid_tier_schedule",
      otherPricing ? "ignored_field" : "skipped_entry",
      ["tiered_pricing"],
      "request-tier pricing schedule is malformed"
    );
    if (!otherPricing) {
      return { pricing: null, diagnostics: [...diagnostics, invalidTier] };
    }
    diagnostics.push(invalidTier);
    delete sanitized.tiered_pricing;
  }

  if (!matchesRawSchema(sanitized)) {
    return { pricing: null, diagnostics: [...diagnostics, schemaDiagnostic(model)] };
  }

  let pricing: ModelPricing;
  try {
    pricing = toModelPricing(model, sanitized, defaultCurrency);
  } catch (error) {
    if (!(error instanceof Error) || error.message !== "Missing supported pricing fields") {
      throw error;
    }
    const hasCostField = Object.keys(sanitized).some((key) => key.includes("cost"));
    return {
      pricing: null,
      diagnostics: [
        ...diagnostics,
        diagnostic(
          model,
          "info",
          hasCostField ? "unsupported_pricing_unit" : "missing_supported_pricing",
          "skipped_entry",
          [],
          hasCostField
            ? "entry uses pricing units that llmcalc does not support"
            : "entry has no supported pricing fields"
        )
      ]
    };
  }
  for (const key of Object.keys(sanitized)) {
    if (unsupportedCostField(key)) {
      diagnostics.push(
        diagnostic(
          model,
          "info",
          "unsupported_pricing_unit",
          "ignored_field",
          [key],
          "pricing field uses a billing unit that llmcalc does not support"
        )
      );
    }
  }
  return { pricing, diagnostics };
}
